raft-compat/rolling_upgrade_test.go

```diff
@@ -0,0 +1,190 @@
+		a, _ := getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+		require.Equal(t, a, leader)
+		getLeader.GetRaft().(*raftprevious.Raft).RemoveServer(raftprevious.ServerID(rLatest.ID(i)), 0, 0)
+		rLatest.Raft(rLatest.ID(i)).(*raftprevious.Raft).Shutdown()
+	}
+	future = getLeader.GetRaft().(*raftprevious.Raft).Apply([]byte("test2"), time.Second)
+	require.NoError(t, future.Error())
+
+	fa := getLeader.GetRaft().(*raftprevious.Raft).AddVoter(raftprevious.ServerID(rUIT.ID(leaderIdx)), raftprevious.ServerAddress(rUIT.Addr(leaderIdx)), 0, 0)
+	utils.WaitFuture(t, fa)
+
+	// Check Leader haven't changed as we haven't replaced it yet
+	a, _ = getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+	require.Equal(t, a, leader)
+	fr := getLeader.GetRaft().(*raftprevious.Raft).RemoveServer(raftprevious.ServerID(rLatest.ID(leaderIdx)), 0, 0)
+	utils.WaitFuture(t, fr)
+	rLatest.Raft(getLeader.GetLocalID()).(*raftprevious.Raft).Shutdown()
+	utils.WaitForNewLeader(t, getLeader.GetLocalID(), rUIT)
+	newLeader := rUIT.GetLeader()
+	require.NotEmpty(t, newLeader)
+	aNew, _ := newLeader.GetRaft().(*raft.Raft).LeaderWithID()
+	require.NotEqual(t, aNew, leader)
+
+	require.Equal(t, newLeader.NumLogs(), 2)
+
+}
+
+// TestRaft_ReplaceUpgrade This test perform a rolling upgrade by removing an old node,
+// and create a new node with the same store until all old nodes are cycled to new nodes.
+// This simulate the advised way of upgrading in Consul.
+func TestRaft_ReplaceUpgrade(t *testing.T) {
+
+	initCount := 3
+	rLatest := testcluster.NewRaftCluster(t, testcluster.InitLatest, initCount, "raftOld")
+	configuration := raftprevious.Configuration{}
+
+	for i := 0; i < initCount; i++ {
+		var err error
+		require.NoError(t, err)
+		configuration.Servers = append(configuration.Servers, raftprevious.Server{
+			ID:      raftprevious.ServerID(rLatest.ID(i)),
+			Address: raftprevious.ServerAddress(rLatest.Addr(i)),
+		})
+	}
+	raft0 := rLatest.Raft(rLatest.ID(0)).(*raftprevious.Raft)
+	boot := raft0.BootstrapCluster(configuration)
+	if err := boot.Error(); err != nil {
+		t.Fatalf("bootstrap err: %v", err)
+	}
+	utils.WaitForNewLeader(t, "", rLatest)
+	getLeader := rLatest.GetLeader()
+	require.NotEmpty(t, getLeader)
+	a, _ := getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+	require.NotEmpty(t, a)
+	future := getLeader.GetRaft().(*raftprevious.Raft).Apply([]byte("test"), time.Second)
+	utils.WaitFuture(t, future)
+
+	leader, _ := getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+	require.NotEmpty(t, leader)
+	// Upgrade all the followers
+	leaderIdx := 0
+	for i := 0; i < initCount; i++ {
+		if getLeader.GetLocalID() == rLatest.ID(i) {
+			leaderIdx = i
+			continue
+		}
+
+		// Check Leader haven't changed
+		a, _ := getLeader.GetRaft().(*raftprevious.Raft).LeaderWithID()
+		require.Equal(t, a, leader)
+
+		//
+		getLeader.GetRaft().(*raftprevious.Raft).RemoveServer(raftprevious.ServerID(rLatest.ID(i)), 0, 0)
```

Comment:
Hmm I'm in two minds about this step... Consul's default mode for servers is to have `leave_on_terminate = false` which I think is correct - generally killing a server is either due to an in-place upgrade or a crash. In either case the intent is not to reduce the size of the quorum so not leaving makes sense.

The confusing part is that our docs since forever have actively recommended to manually perform a `consul leave` on the server before stopping it which I've never really liked for the above reasons. I think the benefit of doing so is that when the leader leaves gracefully leadership can transfeer more smoothly, but we now have Leadership Transfer that should explicitly handle that without messing with the quorum.

I think there was some discussion recently about how leaving first also increases the chances of leader disruption because the node leaves but then still interacts with the other raft nodes somehow but I don't recall the details.

I _think_ it would be a better test if we didn't re-configure raft here as that is closer to how a straight replacement upgrade (without manual leave steps) would work.

Reply:
FWIW those upgrade docs are being changed and in newer versions the user could force a leadership transfer when upgrading the leader instead of performing a leave.

Bad things can happen when a server leaves the cluster but will come back. Mainly that after leaving if raft on the node isn't turned off quickly enough it will attempt to kick off leader elections running up the term. The behavior we observed with a recent issue is that if that node runs up the term by X then when it rejoins it will trigger X leader elections until all other nodes agree on the term. This has the effect of destabilizing the cluster during the upgrade.

Reply:
I think, independently from the upgrade procedure documented in Consul, we should have tests for all of those scenarios as they may happen during upgrade for different reasons and we need to ensure that the cluster end up stable at the end. I can set up a table test that include those scenarios.

@mkeeler The issue you are referring to is related to not having raft prevote implemented. Actually in the test there is an extra call to wait for leadership because of that and I added a comment about it not being necessary when we have prevote.

Reply:
Also that make realize that I'm changing the name of the nodes when upgrading, which do not really reflect an upgrade scenario, I will fix that.